sql/pkg/repository/memory/memory.go

```diff
@@ -0,0 +1,162 @@
+	var slice []model.Contact
+	for row.Next() {
+		var tmp model.Contact
+		err := row.Scan(&tmp.ID, &tmp.FirstName, &tmp.LastName, &tmp.Phone, &tmp.Email)
+		if err != nil {
+			return contact, err
+		}
+
+		slice = append(slice, tmp)
+	}
+
+	for _, c := range slice {
+		if c.Email == contact.Email {
+
+			return model.Contact{}, fmt.Errorf("contact with email %q already exists", c.Email)
+		}
+
+		if c.Phone == contact.Phone {
+			return model.Contact{}, fmt.Errorf("contact with phone %q already exists", c.Phone)
+		}
+	}
+
+	r.lastID++
+	contact.ID = r.lastID
+	_, err = r.storage.Exec("insert into Contacts (ID, FirstName, LastName, Phone, Email) values ($1, $2, $3, $4, $5)", contact.ID, contact.FirstName, contact.LastName, contact.Phone, contact.Email)
+
+	return contact, nil
+}
+
+func (r *ContactsRepositoryInMemory) ListAll() ([]model.Contact, error) {
+	var result []model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contacts")
+	if err != nil {
+		return result, err
+	}
+
+	defer row.Close()
+	for row.Next() {
+		var tmp model.Contact
+		err := row.Scan(&tmp.ID, &tmp.FirstName, &tmp.LastName, &tmp.Phone, &tmp.Email)
+		if err != nil {
+			return result, err
+		}
+
+		result = append(result, tmp)
+	}
+
+	return result, nil
+}
+
+func (r *ContactsRepositoryInMemory) GetByID(id uint) (model.Contact, error) {
+	var contact model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contacts where ID = ?", id)
+	if err != nil {
+		return model.Contact{}, nil
+	}
+
+	defer row.Close()
+	err = row.Scan(&contact.ID, &contact.FirstName, &contact.LastName, &contact.Phone, &contact.Email)
+	if err != nil {
+		return contact, nil
+	}
+
+	return contact, nil
+}
+
+func (r *ContactsRepositoryInMemory) GetByPhone(phone string) (model.Contact, error) {
+	var contact model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contacts where ID = ?", phone)
+	if err != nil {
+		return model.Contact{}, nil
+	}
+
+	defer row.Close()
+	err = row.Scan(&contact.ID, &contact.FirstName, &contact.LastName, &contact.Phone, &contact.Email)
+	if err != nil {
+		return contact, nil
+	}
+
+	return contact, nil
+}
+
+func (r *ContactsRepositoryInMemory) GetByEmail(email string) (model.Contact, error) {
+	var contact model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contact where Email = ?", email)
+	if err != nil {
+		return contact, err
+	}
+
+	defer row.Close()
+	err = row.Scan(&contact.ID, &contact.FirstName, &contact.LastName, &contact.Phone, &contact.Email)
+	if err != nil {
+		return contact, nil
+	}
+
+	return model.Contact{}, fmt.Errorf("record not found")
+}
+
+func (r *ContactsRepositoryInMemory) SearchByName(name string) ([]model.Contact, error) {
+	var contacts []model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contacts where Email = ?", name)
```

Comment:
Search in memory repository allow you to search by prefix, there is strict equal condition see https://www.w3schools.com/sql/sql_like.asp